src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,171 @@
+        {
+            return ProcessProgressNotificationAsync(args.MethodName, args.ParameterToken);
+        }
+
+        private async Task ProcessProgressNotificationAsync(string methodName, JToken parameterToken)
+        {
+            if (methodName != Methods.ProgressNotificationName ||
+               !parameterToken.HasValues ||
+               parameterToken["value"] is null ||
+               parameterToken["token"] is null)
+            {
+                return;
+            }
+
+            var token = parameterToken["token"].ToObject<string>(); // IProgress<object>>();
+
+            if (string.IsNullOrEmpty(token) || !_activeRequests.TryGetValue(token, out var request))
+            {
+                return;
+            }
+
+            var value = parameterToken["value"];
+
+            try
+            {
+                await request.OnProgressResult(value).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Initial handle request has been cancelled
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            var cts = new CancellationTokenSource();
+
+            lock (request.RequestLock)
+            {
+                var existingCTS = request.CancellationTokenSource;
+                if (existingCTS != null &&
+                    existingCTS.Token.CanBeCanceled &&
+                    !existingCTS.Token.IsCancellationRequested)
+                {
+                    existingCTS.Cancel();
+                }
+
+                request.CancellationTokenSource = cts;
+            }
+
+            _ = CompleteAfterDelayAsync(token, request); // Fire and forget
+        }
+
+        private async Task CompleteAfterDelayAsync(string token, ProgressRequest request)
+        {
+            if (request.CancellationTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
+
+            try
+            {
+                await Task.Delay(request.TimeoutAfterLastNotify, request.CancellationTokenSource.Token).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Task cancelled, new progress notification received.
+                // Don't allow handler to return
+                return;
+            }
+
+            ProgressCompleted(token);
+        }
+
+        internal override bool TryListenForProgress(
+            string requestId,
+            Func<JToken, Task> onProgressResult,
+            TimeSpan timeoutAfterLastNotify,
+            out Task onCompleted)
+        {
+            var onCompletedSource = new TaskCompletionSource<bool>();
+            var progressRequest = new ProgressRequest(onProgressResult, timeoutAfterLastNotify, onCompletedSource);
+            onCompleted = onCompletedSource.Task;
```

Comment:
For `out` parameters if the method returns `false` it's convention to have the `out` parameter be its default value (`null` in this case)